### PR TITLE
Benchmark Supported Data Types

### LINKED ISSUE #2200: Supported data types benchmark
### Problem Description

As a developer it would be nice to  ensure that the data types supported by SDV remain functional after any changes to the codebase.

### Expected behavior

Develop a benchmarking suite that automatically tests and asserts the support for each currently supported data type. This suite should ensure that for each data type, the following functionalities are individually tested and verified:

- Metadata Detection:
  - Validate that the data type is correctly identified by SDV.
- Metadata Validation: 
  - Confirm that the metadata validates the given data type (with data).
- Synthesizer Fit:
  - Ensure that the synthesizers can be fit without errors.
- Synthesizer Sample:
  - Verify that the synthesizers can generate samples for the given data type (and returned in this same data type).
- Constraints Fit:
  - Test that constraints related to the data type are correct during fitting process.
- Constraints Sample:
  - Validate that the constraints are respected when generating samples from the synthesizer and the data type is conserved.

### PR DESCRIPTION
**Pull request by pvk-developer**
Resolves #2200 
CU-86b1xxa7d

This pull request introduces a benchmarking suite designed to test all supported data types for our synthesizers and validation processes. Key changes and additions include:

* Benchmarking Integration: Added a new benchmarking framework to evaluate the functionality of all supported data types.

* Private Spreadsheet Integration: The benchmarking results are compared against data read from a private spreadsheet. This spreadsheet contains the expected outcomes for each data type, ensuring that our tests remain accurate and relevant.

* Automated Test Failures: If a data type is no longer supported due to recent changes, the test will automatically fail. This helps in catching unsupported data types and ensures that our system continues to function correctly with all valid data types.

### CODE REVIEW

File: tests/benchmark/supported_dtypes_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,641 @@
...
+            [pd.Timedelta(days=1), pd.Timedelta(days=2), pd.Timedelta(days=3)],
+            dtype='timedelta64[ns]',
+        )
+    }),
+    'pd.Period': pd.DataFrame({
+        'pd.Period': pd.Series(pd.period_range('2023-01', periods=3, freq='M')),
+    }),
+    'pd.Complex': pd.DataFrame({
+        'pd.Complex': pd.Series([1 + 1j, 2 + 2j, 3 + 3j], dtype='complex128'),
+    }),
```
Comment (gsheni) [inline tests/benchmark/supported_dtypes_benchmark.py:157]: Add pyarrow dtypes:
https://pandas.pydata.org/docs/reference/arrays.html#pyarrow
Could be a different dictionary

File: tests/benchmark/supported_dtypes_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,641 @@
...
+        )
+    }),
+    'np.datetime64': pd.DataFrame({
+        'np.datetime64': pd.Series([
+            np.datetime64('2023-01-01T00:00:00'),
+            np.datetime64('2024-01-01T00:00:00'),
+            np.datetime64('2025-01-01T00:00:00'),
+        ])
+    }),
+    'np.timedelta64': pd.DataFrame({
```
Comment (gsheni) [inline tests/benchmark/supported_dtypes_benchmark.py:238]: Add a few more numpy dtypes:
```
import numpy as np

np.dtypes.Float16DType()
np.dtypes.Float32DType())
np.dtypes.Float64DType()

**Review by gsheni — CHANGES_REQUESTED** [review]

File: tests/benchmark/supported_dtypes_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,710 @@
...
+            low_value = data[dtype].min()
+            high_value = data[dtype].max()
+            constraint['constraint_parameters']['low_value'] = low_value
+            constraint['constraint_parameters']['high_value'] = high_value
+
+    elif constraint_class == 'FixedIncrements':
+        if sdtype == 'numerical':
+            values = [10, 20, 30, 40]
+            if dtype.startswith('pd'):
+                values.append(None)
```
Comment (amontanez24) [inline tests/benchmark/supported_dtypes_benchmark.py:524]: should we also put nans in float columns?

File: tests/benchmark/utils.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    """Return previous result for a given ``dtype`` and method."""
+    data = read_excel(BENCHMARK_FILE_ID)
+    python_version = detect_python_version()
+    df = data[python_version]
+    filtered_row = df[df['dtype'] == dtype]
+    value = filtered_row[method].to_numpy()[0]
+    return value
+
+
+def _get_drive_client():
```
Comment (amontanez24) [inline tests/benchmark/utils.py:37]: can we actually pull all the gdrive stuff into a higher level module? Like tests/_external/google_utils or something like that?

File: tests/benchmark/supported_dtypes_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,710 @@
...
+        ])
+    }),
+    'np.timedelta64': pd.DataFrame({
+        'np.timedelta64': pd.Series(
+            [np.timedelta64(1, 'D'), np.timedelta64(2, 'h'), np.timedelta64(3, 'm')],
+        )
+    }),
+}
+
+PYARROW_DTYPES = {
```
[the remark this answers is not in the document]
Comment (gsheni) [reply tests/benchmark/supported_dtypes_benchmark.py:246]: - Starting with the next major pandas release (3.0), pyarrow will be a required dependency of pandas. In addition, the default inferred type will change from `object` to `pyarrow.string.`
  - https://pandas.pydata.org/pdeps/0010-required-pyarrow-dependency.html

- Given this, I wanted to start tracking sdv support for this

[the remark this answers is not in the document]
Comment (pvk-developer) [reply tests/benchmark/supported_dtypes_benchmark.py:246]: Sorry I need to update the spreadsheet to have them and we will be running against them.

File: tests/benchmark/supported_dtypes_benchmark.py (hunk 3 above)
Comment (pvk-developer) [reply tests/benchmark/supported_dtypes_benchmark.py:524]: Yes makes sense, let me add it for all `floats`.

File: tests/benchmark/supported_dtypes_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+    'pa.int8': pd.DataFrame({'pa.int8': pd.Series([1, -1, 127], dtype=pd.ArrowDtype(pa.int8()))}),
```
Comment (gsheni) [inline tests/benchmark/supported_dtypes_benchmark.py:247]: Should there be NaNs for the all the columns? I believe pyarrow supports that

File: tests/benchmark/supported_dtypes_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+
+
+def _create_multi_column_constraint_data_and_metadata(constraint, data, dtype, sdtype, metadata):
+    _dtype = data.dtypes[dtype]
+    constraint_class = constraint.get('constraint_class')
+    constraints = []
+    if constraint_class == 'FixedCombinations':
+        for dtype_name, dtype_data in {**PANDAS_DTYPES, **NUMPY_DTYPES}.items():
+            dtype_sdtype = EXPECTED_METADATA_SDTYPES.get(dtype_name, 'unknown')
+            if dtype_sdtype in ('categorical', 'boolean'):
```
Comment (amontanez24) [inline tests/benchmark/supported_dtypes_benchmark.py:387]: I think int columns can be sdtype categorical and are valid to be used in these constraints

File: .github/workflows/dtypes_benchmark.yml [hunk 8]
```diff
@@ -0,0 +1,77 @@
+name: Data Types Benchmark
+
+on:
+  push:
+    branches:
+      - main
```
[the remark this answers is not in the document]
Comment (pvk-developer) [reply .github/workflows/dtypes_benchmark.yml:6]: Yes, the benchmark is only for `message` in slack and `updating the gdrive`.

File: .github/workflows/integration.yml [hunk 9]
```diff
@@ -11,7 +11,7 @@ jobs:
     strategy:
       fail-fast: true
       matrix:
-        python-version: [ '3.8', '3.9', '3.10', '3.11', '3.12']
+        python-version: ['3.8', '3.12']
```
Comment (amontanez24) [inline .github/workflows/integration.yml:14]: why did we change this?

File: tests/_external/slack_utils.py [hunk 10]
```diff
@@ -34,9 +34,57 @@ def post_slack_message(channel, text):
...
+
+def update_message(channel, new_text, message_ts):
+    """Update an existing message in a Slack channel.
+
+    Args:
+        channel (str):
+            The name of the channel where the message is located.
+        message_ts (str):
+            The timestamp of the message to be edited.
+        new_text (str):
```
Comment (amontanez24) [inline tests/_external/slack_utils.py:76]: we should match the order

**Review by amontanez24 — COMMENTED** [review]
Looking good!

File: tests/benchmark/supported_dtypes_benchmark.py (hunk 2 above)
Comment (pvk-developer) [reply tests/benchmark/supported_dtypes_benchmark.py]: This are the default ones already when running `NumPy`

**Review by amontanez24 — APPROVED** [review]

File: .github/workflows/dtypes_benchmark.yml [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  run_dtypes_benchmark:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
```
Comment (gsheni) [inline .github/workflows/dtypes_benchmark.yml:18]:
```suggestion
        uses: actions/checkout@v4
```

File: .github/workflows/dtypes_benchmark.yml [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+          name: results-${{ matrix.python-version }}
+          path: results/*.json
+
+  generate_dtypes_report:
+    runs-on: ubuntu-latest
+    needs: run_dtypes_benchmark
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
```
Comment (gsheni) [inline .github/workflows/dtypes_benchmark.yml:55]:
```suggestion
        uses: actions/checkout@v4
```

**Review by gsheni — APPROVED** [review]

**Pull request merged**